android/src/main/java/com/reactnativecommunity/asyncstorage/AsyncStorageExpoMigration.java

```diff
@@ -0,0 +1,154 @@
+package com.reactnativecommunity.asyncstorage;
+
+import android.content.Context;
+import android.os.Build;
+import android.util.Log;
+
+import androidx.annotation.RequiresApi;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+
+// A utility class that migrates a scoped AsyncStorage database to RKStorage.
+// This utility only runs if the RKStorage file has not been created yet.
+public class AsyncStorageExpoMigration {
+    static final String LOG_TAG = "AsyncStorageExpoMigration";
+
+    public static void migrate(Context context) {
+        // Only migrate if the default async storage file does not exist.
+        if (isAsyncStorageDatabaseCreated(context)) {
+            return;
+        }
+
+        ArrayList<File> expoDatabases = getExpoDatabases(context);
+
+        File expoDatabase = getLastModifiedFile(expoDatabases);
+
+        if (expoDatabase == null) {
+            Log.v(LOG_TAG, "No scoped database found");
+            return;
+        }
+
+        try {
+            // Create the storage file
+            ReactDatabaseSupplier.getInstance(context).get();
+            copyFile(new FileInputStream(expoDatabase), new FileOutputStream(context.getDatabasePath(ReactDatabaseSupplier.DATABASE_NAME)));
+            Log.v(LOG_TAG, "Migrated most recently modified database " + expoDatabase.getName() + " to RKStorage");
+        } catch (Exception e) {
+            Log.v(LOG_TAG, "Failed to migrate scoped database " + expoDatabase.getName());
+            e.printStackTrace();
+            return;
+        }
+
+        try {
+            for (File file : expoDatabases) {
+                if (file.delete()) {
+                    Log.v(LOG_TAG, "Deleted scoped database " + file.getName());
+                } else {
+                    Log.v(LOG_TAG, "Failed to delete scoped database " + file.getName());
+                }
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        Log.v(LOG_TAG, "Completed the scoped AsyncStorage migration");
+    }
+
+    private static boolean isAsyncStorageDatabaseCreated(Context context) {
+        return context.getDatabasePath(ReactDatabaseSupplier.DATABASE_NAME).exists();
+    }
+
+    // Find all database files that the user may have created while using Expo.
+    private static ArrayList<File> getExpoDatabases(Context context) {
+        ArrayList<File> scopedDatabases = new ArrayList<>();
+        try {
+            File databaseDirectory = context.getDatabasePath("noop").getParentFile();
```

Comment:
Are we passing in `noop` because it doesn't matter and we just want the parent path? I.e. Android doesn't actually check whether a database named "noop" exists?

Reply:
correct